server/src/main/java/org/elasticsearch/action/search/SearchPhaseController.java

```diff
@@ -247,8 +253,49 @@ TopDocs mergeTopDocs(Collection<TopDocs> results, int topN, int from) {
             mergedTopDocs = CollapseTopFieldDocs.merge(sort, from, topN, shardTopDocs, setShardIndex);
         } else if (topDocs instanceof TopFieldDocs) {
             TopFieldDocs firstTopDocs = (TopFieldDocs) topDocs;
-            final Sort sort = new Sort(firstTopDocs.fields);
+            Sort sort = new Sort(firstTopDocs.fields);
             final TopFieldDocs[] shardTopDocs = results.toArray(new TopFieldDocs[numShards]);
+
+            int sortLength = shardTopDocs[0].fields.length;
+            for (int sortIdx = 0; sortIdx < sortLength; sortIdx++) {
+                final int idx = sortIdx;
+
+                boolean containsNanosecond = Arrays.stream(shardTopDocs)
+                    .filter(d -> d.totalHits.value > 0)
+                    .map(d -> d.fields[idx].getClass().getName())
+                    .anyMatch(className -> SortedNanosecondsNumericSortField.class.getName().equals(className));
```

Comment:
Do we need to map to the class name ? You can use `.getClass().equals(SortedNanosecondsNumericSortField.class)` directly ?

Reply:
fixed